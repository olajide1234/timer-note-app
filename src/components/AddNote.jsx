import React, { useState } from 'react'
import PropTypes from 'prop-types';
import { Button, Form, Input, } from 'antd';
import { useDispatch } from "react-redux";
const { TextArea } = Input;

/**
 * Component for adding notes
 * 
 * @component
 * @example
 * const time = '12:12:12' 
 * const stopFn = f => f
 * currentTimerStopId = 2
 * return (
 *   <AddNote time={time} stopFn={stopFn} currentTimerStopId=2  />
 * ) 
 */
export function AddNote(props) {

  const dispatch = useDispatch();
  const { time, stopFn, currentTimerStopId } = props;
  const [note, setNote] = useState('');
  const [validate, setValidate] = useState({ status: '', help: '' });

  /**
   * Handles submission of the note
   * 
   * @param   {object} event Payload from the form
   * @returns  {void}
   */
  const handleSubmit = (event) => {
    event.preventDefault();
    if (!note) {
      setValidate({ status: 'error', help: 'Note can not be empty' });
      return;
    }
    stopFn(currentTimerStopId);
    dispatch({ type: 'ADD_NOTE', payload: { time, text: note } });
    setNote('');
  }

  return (
    <div>
      <div className='headerText'>Add Notes</div>
      <div className='smallTopMargin'>
        <Form onSubmit={handleSubmit}>
          <Form.Item
            validateStatus={validate.status}
            help={validate.help}
          >
            <TextArea
              placeholder='Jot your note here'
              rows='5'
              cols='33'
              value={note}
              allowClear
              onChange={(e) => {
                if (validate.status === 'error') {
                  setValidate({ status: '', help: '' })
                }
                setNote(e.target.value)
              }}
            />
          </Form.Item>
          <Button type='primary' htmlType='submit'>
            Save note
          </Button>
        </Form>
      </div>
    </div>
  );
}

AddNote.propTypes = {
  /**
   * Current timer value for note
   */
  time: PropTypes.string.isRequired,

  /**
   * Function to stop timer
   */
  stopFn: PropTypes.func.isRequired,

  /**
   * Id of current timer to be stopped
   */
  currentTimerStopId: PropTypes.number.isRequired
}

