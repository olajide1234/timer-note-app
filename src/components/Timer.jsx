import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button, Row, Col } from 'antd';

/**
 * Timer component view
 *
 * @component
 * @example
 * const time = '12:12:12'
 * const startFn = f => f
 * const resetFn = f => f
 * const stopFn = f => f
 * const currentTimerStopId = 2
 * const timerPause = true
 * return (
 *   <Timer time={time} startFn={startFn} resetFn={resetFn} stopFn={stopFn} currentTimerStopId={currentTimerStopId} timerPause={timerPause}  />
 * )
 */
export function Timer(props) {

  const { time, startFn, resetFn, stopFn, currentTimerStopId, timerPause } = props;
  const [buttonDisable, setButtonDisable] = useState({ start: false, pause: false });

  return (
    <div>
      <div className='headerText'>Control Timer</div>
      <div className='timerFace'>
        <div>{time}</div>
      </div>
      <Row type="flex" justify="space-between">
        <Col>
          <Button
            disabled={timerPause ? false : buttonDisable.start}
            onClick={() => { startFn(); setButtonDisable({ start: true, pause: false }) }}
          >
            Start
          </Button>
        </Col>
        <Col>
          <Button
            disabled={buttonDisable.pause || timerPause}
            onClick={() => { stopFn(currentTimerStopId); setButtonDisable({ start: false, pause: true }) }}
          >
            Pause
          </Button>
        </Col>
        <Col>
          <Button
            disabled={buttonDisable.reset}
            onClick={() => { resetFn(currentTimerStopId); setButtonDisable({ start: false, pause: false }) }}
          >
            Reset
          </Button>
        </Col>
      </Row>
    </div>
  );
}

Timer.propTypes = {
  /**
   * Current timer value to display
   */
  time: PropTypes.string.isRequired,

  /**
   * Function to start timer
   */
  startFn: PropTypes.func.isRequired,

  /**
   * Function to reset timer
   */
  resetFn: PropTypes.func.isRequired,

  /**
   * Function to stop timer
   */
  stopFn: PropTypes.func.isRequired,

  /**
   * Id to stop timeout function
   */
  currentTimerStopId: PropTypes.number.isRequired,

  /**
   * Bool to show if timer is paused
   */
  timerPause: PropTypes.bool.isRequired
}
