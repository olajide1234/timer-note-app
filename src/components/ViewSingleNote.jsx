import React from 'react'
import PropTypes from 'prop-types';
import { Card, Button, Modal } from 'antd';
const { confirm } = Modal;

/**
 * Single note view
 *
 * @component
 * @example
 * const id = 2
 * const time = '12:12:12'
 * const text = 'Sample text'
 * const deleteFn = f => f
 * return (
 *   <ViewSingleNote id={id} time={time} text={text} deleteFn={deleteFn} />
 * )
 */
export function ViewSingleNote(props) {

  const { id, time, text, deleteFn } = props;

  /**
   * Handles showing modal for confirmation
   *
   * @param   {string} note Content of note to be deleted
   * @param   {int} id Id of note to be deleted
   * @param   {function} deleteFn Function to call for delete
   * @returns  {void}
   */
  const showConfirm = (note, id, deleteFn) => {
    confirm({
      title: 'Are you sure you wish to permanently delete the following note ?',
      content: note,
      onOk() {
        deleteFn(id)
      },
      onCancel() { },
    });
  }

  return (
    <div className='smallerTopMargin'>
      <Card
        size="small"
        title={time}
        extra={
          <Button
            type="danger"
            size="small"
            onClick={() => showConfirm(text, id, deleteFn)}
          >
            Delete
          </Button>
        }
        style={{ width: 250 }}
      >
        <p>{text}</p>
      </Card>
    </div>
  );
}

ViewSingleNote.propTypes = {
  /**
   * Id of note to be deleted
   */
  id: PropTypes.number.isRequired,

  /**
   * Note timestamp
   */
  time: PropTypes.string.isRequired,

  /**
   * Note text content
   */
  text: PropTypes.string.isRequired,

  /**
   * Function to call for note delete
   */
  deleteFn: PropTypes.func.isRequired,
}
