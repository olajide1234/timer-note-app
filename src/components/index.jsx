export { ViewSingleNote } from './ViewSingleNote';
export { Timer } from './Timer';
export { AddNote } from './AddNote';
