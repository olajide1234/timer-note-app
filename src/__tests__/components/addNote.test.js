import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { AddNote } from '../../components';
import renderer from 'react-test-renderer';

const mockStore = configureMockStore();

it('should render component correctly', () => {
  const store = mockStore([{ id: 1, time: '12:12:12', text: 'Here is a sample note. Add more notes to see your note above this.' }]);
  const tree = renderer.create(
    <Provider store={store}>
      <AddNote time='11:11:11' />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should not add note due to empty note submission', () => {
  const store = mockStore([{ id: 1, time: '12:12:12', text: 'Here is a sample note. Add more notes to see your note above this.' }]);
  const wrapper = mount(
    <Provider store={store}>
      <AddNote time='11:11:11' />
    </Provider>
  );
  wrapper.find('button').at(0).simulate('submit');
  const action = store.getActions();
  expect(action).toEqual([]);
});

it('should add note successfully', () => {
  const stopFn = jest.fn();
  const store = mockStore([{ id: 1, time: '12:12:12', text: 'Here is a sample note. Add more notes to see your note above this.' }]);
  const wrapper = mount(
    <Provider store={store}>
      <AddNote time='11:11:11' stopFn={stopFn} currentTimerStopId='2' />
    </Provider>
  );
  wrapper.find('textarea').simulate("change", { target: { value: "foo" } });
  wrapper.find('button').at(0).simulate('submit');
  wrapper.update();
  const action = store.getActions();
  expect(action).toEqual([{ type: 'ADD_NOTE', payload: { time: '11:11:11', text: 'foo' } }]);
});
