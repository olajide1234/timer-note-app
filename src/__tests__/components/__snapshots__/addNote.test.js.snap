// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly 1`] = `
<div>
  <div
    className="headerText"
  >
    Add Notes
  </div>
  <div
    className="smallTopMargin"
  >
    <form
      className="ant-form ant-form-horizontal"
      onSubmit={[Function]}
    >
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control-wrapper"
          style={Object {}}
        >
          <div
            className="ant-form-item-control"
          >
            <span
              className="ant-form-item-children"
            >
              <span
                className="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
              >
                <textarea
                  className="ant-input"
                  cols="33"
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Jot your note here"
                  rows="5"
                  style={Object {}}
                  value=""
                />
              </span>
            </span>
          </div>
        </div>
      </div>
      <button
        className="ant-btn ant-btn-primary"
        onClick={[Function]}
        type="submit"
      >
        <span>
          Save note
        </span>
      </button>
    </form>
  </div>
</div>
`;
